contract/src/main.rs

```diff
@@ -16,193 +19,139 @@ use types::{
-        "store_tuple2",
-        CLType::Tuple2([Box::new(CLType::String), Box::new(CLType::U512)]),
+    entry_points.add_entry_point(EntryPoint::new(
+        String::from("delete"),
+        vec![Parameter::new("uref", CLType::String)],
+        CLType::Unit,
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
     ));
-
     let (contract_hash, _) =
         storage::add_contract_version(contract_package_hash, entry_points, Default::default());
-    runtime::put_key("kvstorage_contract", contract_hash.into());
+    runtime::put_key("kvstorage_contract_hash", contract_hash.into());
     let contract_hash_pack = storage::new_uref(contract_hash);
-    runtime::put_key("kvstorage_contract_hash", contract_hash_pack.into())
+    runtime::put_key("kvstorage_contract_hash_wrapped", contract_hash_pack.into())
 }
 
-fn read_and_store<T: CLTyped + FromBytes + ToBytes>() {
-    let name: String = runtime::get_named_arg("name");
-    let value: T = runtime::get_named_arg("value");
-    set_key(name.as_str(), value);
+fn create_key<T: ToBytes + CLTyped>(name: &str, dict_name: &str, dict_value: T) {
+    match runtime::get_key(name) {
+        None => {
+            // storage::new_dictionary function creates new URef that represents a seed for a dictionary partition
+            // of the global state and puts it under named keys.
+            // Any value can be stored directly under named keys using storage::new_uref function.
+            // But for more organisational structure, it's better to store values under dictionary_item_key in the
+            // dictionary and control this dictionary with one dictionary_seed_uref.
+            // It's possible to have multiple dictionaries for big data structure.
+            let uref = storage::new_dictionary(dict_name).unwrap_or_revert();
+            storage::dictionary_put(uref, dict_name, dict_value);
+            runtime::put_key(name, uref.into());
+        }
+        Some(_) => {
+            runtime::revert(ApiError::None);
+        }
+    }
 }
 
-fn endpoint(name: &str, value_type: CLType) -> EntryPoint {
-    EntryPoint::new(
-        String::from(name),
-        vec![
-            Parameter::new("name", CLType::String),
-            Parameter::new("value", value_type),
-        ],
-        CLType::Unit,
-        EntryPointAccess::Public,
-        EntryPointType::Contract,
-    )
+fn read_key<T: FromBytes + CLTyped>(name: &str, dict_name: &str) -> T {
+    match runtime::get_key(name) {
+        Some(key) => match key {
+            Key::URef(uref) => {
+                let result = storage::dictionary_get::<T>(uref, dict_name).unwrap_or_revert();
+                result.unwrap()
+            }
+            _ => runtime::revert(ApiError::None),
+        },
+        None => {
+            runtime::revert(ApiError::None);
+        }
+    }
 }
 
-fn set_key<T: ToBytes + CLTyped>(name: &str, value: T) {
+fn update_key<T: ToBytes + CLTyped>(name: &str, dict_name: &str, dict_value: T) {
```

Comment:
You only allow String in the entry point so there is no need to make any of the function arguments generic